### PR TITLE
Change class variables to thread local variables

### PR DESCRIPTION
**Pull request by anirudh2290**
## Description ##

There are many places where class variables which end up modifying the state from multiple threads and causing issues like the one here: https://github.com/apache/incubator-mxnet/issues/9920 .
This PR is currently WIP because  more investigation is required on  `test_multi_worker_forked_data_loader` and `test_laop_3`(able to reproduce this test failure without my changes also). Also, more tests need to be added.

@piiswrong 

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: python/mxnet/context.py [hunk 1]
```diff
@@ -60,18 +62,16 @@ class Context(object):
     >>> gpu_array.context
     gpu(1)
     """
-    # static class variable
-    default_ctx = None
```
Comment (szha) [inline python/mxnet/context.py]: This will break a lot of existing code. Is this change intended for 2.0?

Comment (piiswrong) [reply python/mxnet/context.py]: I don't think this was meant to be public API. current_context() is used for accessing this.

Comment (anirudh2290) [reply python/mxnet/context.py]: @piiswrong told me that default_ctx, devtype2str and devstr2type where not meant to be public API. This will also be renamed to _default_ctx. Users are expected to use `current_context` or `default_context` in test_utils. I agree though that this will break some existing code. I was of the opinion that the release after 1.2 is 2.0 but I am not sure.

File: python/mxnet/base.py [hunk 2]
```diff
@@ -98,6 +99,66 @@ class MXCallbackList(ctypes.Structure):
...
+
+        def __enter__(self):
+            self.save_local_var()
+            self.set_local_var()
+            return self
+
+        def __exit__(self, ptype, value, trace):
+            self.load_local_var()
+else:
+    class ThreadLocalRegistry(object, metaclass=ThreadLocalRegistryMeta):
```
Comment (piiswrong) [inline python/mxnet/base.py]: This looks really complicated. why do you have to use inheritance? Isn't composition better?

File: python/mxnet/base.py [hunk 3]
```diff
@@ -98,6 +99,66 @@ class MXCallbackList(ctypes.Structure):
...
+        """Structure that holds threadlocal variables and sets context manager"""
+
+        def __init__(self, name):
```
Comment (piiswrong) [inline python/mxnet/base.py]: why do you need a name?

File: python/mxnet/context.py (hunk 1 above)
Comment (piiswrong) [reply python/mxnet/context.py]: If there are users who depend on this, we can make a backward compatible wrapper with classmethod and property:https://stackoverflow.com/questions/1697501/python-staticmethod-with-property

File: python/mxnet/base.py (hunk 2 above)
Comment (anirudh2290) [reply python/mxnet/base.py]: Inheritance was used since the thread local variables need to still be bound to the class rather than the object.

File: python/mxnet/base.py (hunk 3 above)
Comment (anirudh2290) [reply python/mxnet/base.py]: The name here is the name of the threadlocal variable for example `default_ctx`. 
I will add an empty or None check here, to prevent from being misused.

File: python/mxnet/attribute.py [hunk 4]
```diff
@@ -31,9 +32,9 @@ class AttrScope(object):
     kwargs
         The attributes to set for all symbol creations in the scope.
     """
-    current = None
```
Comment (piiswrong) [inline python/mxnet/attribute.py]: _current = threading.local()

File: python/mxnet/attribute.py [hunk 5]
```diff
@@ -64,15 +65,15 @@ def get(self, attr):
 
     def __enter__(self):
         # pylint: disable=protected-access
-        self._old_scope = AttrScope.current
```
Comment (piiswrong) [inline python/mxnet/attribute.py]: self._old_scope = AttrScope.current.value

File: python/mxnet/attribute.py [hunk 6]
```diff
@@ -64,15 +65,15 @@ def get(self, attr):
...
+        attr = AttrScope.local.current._attr.copy()
         attr.update(self._attr)
         self._attr = attr
-        AttrScope.current = self
+        self.set_local_var()
         return self
 
     def __exit__(self, ptype, value, trace):
-        assert self._old_scope
-        AttrScope.current = self._old_scope
```
Comment (piiswrong) [inline python/mxnet/attribute.py]: AttrScope.current.value = self._old_scope

File: python/mxnet/context.py (hunk 1 above)
Comment (piiswrong) [reply python/mxnet/context.py]:
```
_default_ctx = threading.local()
@property
@classmethod
def default_ctx():
   return _default_ctx.value
```

File: python/mxnet/context.py [hunk 7]
```diff
@@ -109,15 +112,37 @@ def __repr__(self):
         return self.__str__()
 
     def __enter__(self):
-        self._old_ctx = Context.default_ctx
-        Context.default_ctx = self
+        if not hasattr(Context._default_ctx, "value"):
```
Comment (piiswrong) [inline python/mxnet/context.py]: why do you need this? Isn't it set with https://github.com/apache/incubator-mxnet/pull/10833/files#diff-1a5e06031378f44204fd1da1fffc0b07R145

File: python/mxnet/gluon/block.py [hunk 8]
```diff
@@ -46,10 +47,10 @@ def __init__(self, block):
     @staticmethod
     def create(prefix, params, hint):
         """Creates prefix and params for new `Block`."""
-        current = _BlockScope._current
+        current = getattr(_BlockScope._current, "value", None)
```
Comment (piiswrong) [inline python/mxnet/gluon/block.py]: initialize _BlockScope._current.value to None globally and use _BlockScope._current.value directly here?

File: python/mxnet/name.py [hunk 9]
```diff
@@ -62,14 +65,30 @@ def get(self, name, hint):
         return name
 
     def __enter__(self):
-        self._old_manager = NameManager.current
-        NameManager.current = self
+        if not hasattr(NameManager._current, "value"):
```
Comment (piiswrong) [inline python/mxnet/name.py]: same

File: python/mxnet/name.py [hunk 10]
```diff
@@ -62,14 +65,30 @@ def get(self, name, hint):
...
-
+        NameManager._current.value = self._old_manager
+
+    #pylint: disable=no-self-argument
+    @classproperty
+    def current(cls):
+        warnings.warn("NameManager.current has been deprecated. "
+                      "It is advised to use the `with` statement with NameManager.",
+                      DeprecationWarning)
+        if not hasattr(NameManager._current, "value"):
```
Comment (piiswrong) [inline python/mxnet/name.py]: same

File: python/mxnet/ndarray/ndarray.py [hunk 11]
```diff
@@ -2261,7 +2261,7 @@ def ones(shape, ctx=None, dtype=None, **kwargs):
         The shape of the empty array.
     ctx : Context, optional
         An optional device context.
-        Defaults to the current default context (``mxnet.Context.default_ctx``).
+        Defaults to the current default context (``mxnet.Context._default_ctx.value``).
```
Comment (piiswrong) [inline python/mxnet/ndarray/ndarray.py]: This is public API. change to mxnet.current_context

File: python/mxnet/attribute.py [hunk 12]
```diff
@@ -64,15 +66,35 @@ def get(self, attr):
 
     def __enter__(self):
         # pylint: disable=protected-access
-        self._old_scope = AttrScope.current
-        attr = AttrScope.current._attr.copy()
+        if not hasattr(AttrScope._current, "value"):
```
Comment (piiswrong) [inline python/mxnet/attribute.py]: same

**Comment by piiswrong** [comment]
Please remove the hasatter/getattr statements. Otherwise LGTM

File: python/mxnet/context.py (hunk 7 above)
Comment (anirudh2290) [reply python/mxnet/context.py]: what about a situation like below:
```
import threading
import mxnet as mx

def test_context():
    ctx_list = []
    def f():
        mx.context.Context.default_ctx
        ctx_list.append(Context.default_ctx)
    thread = threading.Thread(target=f)
    thread.start()
    thread.join()

if __name__ == '__main__':
    test_context()
```

Without hasattr this doesn't work.

**Comment by anirudh2290** [comment]
Recording some of the offline conversation: Initially the inheritance and metaclasses was for allowing code reuse of thread local code easily for multiple modules. This code was a bit complicated to understand so we moved the `threading.local` calls to module where it was needed instead of using inheritance and metaclasses for threadlocal. For adding backward compatibility, we have to make `default_ctx` and equivalents in other modules class properties. This required adding logic in base.py for the same. This included adding a metaclass and since metaclass syntax is not the same between python2 and python3 needed to use six. We cannot add six as a dependency, because that restricts users more, and thus we copied the metaclass specific code from six into base.py.

**Comment by cclauss** [comment]
> We cannot add __six__ as a dependency

@szha Is this true?  I find numerous instances where Python2-only builtins like basestring, long, reload, unicode, xrange are used.  Do we need try/except blocks to define tests or can we use __six.moves__?

**Comment by szha** [comment]
@cclauss I'd probably describe this as "six is not necessary", as we already handle most of the import compatibility in base.py where we use the python version to decide the definition, and we've been careful about not using version-specific features in our code base (e.g. we have both py2 and py3 in CI). Thus, adding six would mostly just benefit code that's outside of our code base, such as the examples, in which case the dependency is not managed by mxnet and authors are free to choose.
